File: draft_design_pattern/prototype.py
# prototypeパターンは、既存のオブジェクトを「クローン」することでオブジェクトを作成します。
import copy


class Sheep:
    def __init__(self, name: str, category: str = "オオツノヒツジ") -> None:
        self.name = name
        self.category = category

    def setName(self, name: str) -> None:
        self.name = name

    def getName(self) -> str:
        return self.name

    def setCategory(self, category: str) -> None:
        self.category = category

    def getCategory(self) -> str:
        return self.category


original = Sheep('ジョリー')
print(original.getName())
print(original.getCategory())
print(id(original))

# 参照
reference = original
print(id(reference))

# クローン後、必要なものを変更する
cloned = copy.deepcopy(original)
cloned.setName('ドリー')
print(cloned.getName())
print(cloned.getCategory())
print(id(cloned))

# オリジナル再確認
print(original.getName())

File: draft_design_pattern/prototype.nim
type Sheep = ref object
  name: string
  category: string


proc newSheep(name: string, category: string = "オオツノヒツジ"): Sheep =
  Sheep(name: name, category: category)

proc setName(self: Sheep, name: string) =
  self.name = name

proc getName(self: Sheep): string =
  self.name

proc setCategory(self: Sheep, category: string) =
  self.category = category

proc getCategory(self: Sheep): string =
  self.category


proc main() =
  var original = newSheep("ジョリー")
  echo original.getName()
  echo original.getCategory()

  echo "----"

  # 参照
  var reference = original
  reference.setName("ジョリー参照")
  echo reference.getName()
  echo original.getName()

  echo "----"

  # コピー
  var cloned: Sheep
  cloned.deepCopy(original)

  cloned.setName("ジョリーコピー")
  cloned.setCategory("ヤギ")

  echo cloned.getName()
  echo cloned.getCategory()

  echo "---- オリジナル"

  echo original.getName()
  echo original.getCategory()


main()
